### ISSUE TITLE
Do a better job of detecting compatible game versions for add-ons without `package.json`

### ISSUE THREAD

**Issue by layday** [layday/github-wow-addon-catalogue#3]
Probably from reading the TOC file.

**Comment by torkus** [comment]
I have an open issue right now installing Questie from github. One of the problems I found was that Questie is using a dummy .toc file: https://github.com/Questie/Questie/blob/master/Questie.toc

So I can test for the presence of one or more .toc files to guess supported game versions but I won't really know until I download and parse the contents. Hopefully it's only Questie being weird.

**Comment by layday** [comment]
For add-ons above a certain size (say, 500 KB), could we do something like, perform an HTTP range request to read the zip directory from the end of the add-on zip followed by another range request to read the toc file?

**Comment by torkus** [comment]
that's really interesting, I'd never considered peeking inside a partial zip file. I'm not sure what it's specification is like. Is the index kept at the end of the file? 

Looking at the [wiki page](https://en.wikipedia.org/wiki/ZIP_(file_format)#Structure) it seems like if you could read the file in reverse from the end of the file until the end of the index you could avoid downloading the whole thing, but if all we have is a HTTP range to work with I'm not sure how you'd know where to start downloading from.

Could be a fun investigation.

**Comment by layday** [comment]
Without knowing a whole lot about zip files or range requests, this is how I would approach it:

1. Do a head request to grab the content length
2. Do a range request where the range is from (content length - ~1 KB or however large the directory typically is) to content length
   1. Repeat until you find the end of directory marker
3. Locate the toc file(s) in the directory
   1. If the filenames are [game version-specific](https://github.com/layday/instawow/blob/06cdfc8c626f1e19180e21634ce06900eaace560/src/instawow/matchers.py#L26-L46), stop here
5. Extract the offset and length of the main toc file from the zip directory
6. Do a range request where the range is from offset to (offset + length)
7. Drop the local header and parse the remainder of the payload as text

**Comment by layday** [comment]
I took a stab at it in https://github.com/layday/instawow/commit/2f35ca6b1c1f9e75ccbaff3fa9cb572a70611eb1 without reading the TOC contents.

**Comment by torkus** [comment]
Ah, very clever. I took a quick skim. Are you using connection pooling with that `web_client` I see there? It would make the worse case scenario (a requests for each offset) not so bad.

I discovered early on in strongbox that zip files can be pretty robust as I tried to corrupt the files by cutting chunks out of them to get tests to fail.

**Comment by layday** [comment]
Yeah, they are pooled. I only tried it with Questie so far which has a directory between 25 and 50 KB in size. We could use this project to gather some metrics.

I did also try inserting the toc file at the correct offset in the byte stream but the CRC would fail and Python refused to extract it.

Sent with [ProtonMail](https://protonmail.com/) secure email.

------- Original Message -------

**Comment by layday** [comment]
I figured out what I was doing wrong with the TOC file and have got a working prototype in https://github.com/layday/instawow/compare/ebcd262...1a8b7d8.

**Comment by torkus** [comment]
Not sure if this is related to the work described here but it falls under the ticket title.

I'm looking at the changes to the addons and in a few cases we've lost game versions.

* https://github.com/SnaBe/BetterZoneStats, 'retail' => nil, should be vanilla
* https://github.com/2CDeez/Buvvs, 'retail' => nil, should be unparseable as interface version is weird and it has no packages
* https://github.com/zerasul/ellidice, 'retail' => nil, should be retail
* https://github.com/Cilraaz/EmoteLDB, 'retail' => nil, should be retail
* https://github.com/dendanskealice/farmlog, 'retail' => nil, should be retail (cata)
* https://github.com/naschorr/Of-The-Statistician, 'retail' => nil, should be retail (legion)
* https://github.com/solinisvc/Whisper-Filter-Continued, 'retail' => nil, should be vanilla

Some of these are obvious outliers

**Comment by layday** [comment]
I’m not on my computer to check right now but I believe the reason some of these have become nil is that the addon files are not placed inside a folder in the zip.

Sent with [ProtonMail](https://protonmail.com/) secure email.

------- Original Message -------

**Comment by layday** [comment]
Okay, I've taken a closer look at each one of these:

* https://github.com/SnaBe/BetterZoneStats
  * folder name different, contains version number: `BetterZoneStats-v1.0/BetterZoneStats.toc`
    * add-on won't load
* https://github.com/2CDeez/Buvvs
  * invalid `Interface` version: `113`
    * add-on will still load
* https://github.com/zerasul/ellidice
  * different capitalisation: `ellidice/ElliDice.toc`
    * add-on might load on case-insensitive systems or if the WoW client normalises add-on names, haven't checked
* https://github.com/Cilraaz/EmoteLDB
  * add-on files in zip root
    * add-on won't load
* https://github.com/dendanskealice/farmlog 
  * as above
    * add-on won't load
* https://github.com/naschorr/Of-The-Statistician
  * folder name different, contains version number
    * add-on won't load
* https://github.com/solinisvc/Whisper-Filter-Continued
  * folder name different from TOC name: `WhisperFilter-Continued/WhisperFilter-Cont.toc`
    * add-on won't load

**Comment by torkus** [comment]
When you say the addon will or won't load, are you actually booting WoW and seeing if it will load? Or does WoW have a list of rules somewhere? I've always found the game to be pretty flexible about what it will load (not an addon author though, I don't really know for sure).

**Comment by layday** [comment]
I didn't check, but it used to be the folder and TOC names had to match. I could try them all out.

**Comment by torkus** [comment]
I'm not fussed, I think you're doing a great job. There will always be outliers.

**Comment by layday** [comment]
Yeah, it still works that way. I just checked with BetterZoneStats, Of-The-Statistician and Whisper-Filter-Continued - none of them work.

**Comment by torkus** [comment]
I was just going through this week's diff and noticed a few changes:

lost full game-track-list:
- https://github.com/BreakBB/ExtendedCharacterStats
- should be 'retail', 'classic', 'classic-tbc'

game-track-list changed from 'retail' to 'classic':
- https://github.com/28/NoteworthyII
- should be 'retail' and 'classic'

game-track-list changed from 'classic-tbc' and 'classic' to 'retail':
- https://github.com/sfmict/LibSFDropDown
- should be 'retail', 'classic', 'classic-tbc'

again, not asking you to fix or even investigate, but you might be curious.

**Comment by layday** [comment]
Thanks, ~~these~~ the first one should be fixed now.  [pfQuest](https://github.com/shagu/pfQuest) is still being parsed incorrectly, something to do with all the different language versions I presume.

**Comment by layday** [comment]
[NoteworthyII](https://github.com/28/NoteworthyII) has separate releases for retail and classic which we don't account for. I suppose we could look for keywords like "retail" and "classic" in the tag name of the *x* latest releases.

**Comment by layday** [comment]
[LibSFDropDown](https://github.com/sfmict/LibSFDropDown) wasn't returned by the search API which is why it wasn't updated just now, but it should be fixed.

**Comment by torkus** [comment]
> I suppose we could look for keywords like "retail" and "classic" in the tag name of the x latest releases.

Strongbox does this, it's not such a great idea in my opinion. Questie in particular has been adding classic support, taking it away, adding it again, creating a 'universal' addon, taking it away ... past support of a game version is no indication of support in any specific release.

**Comment by layday** [comment]
There are some popular add-ons that have got separate releases per flavour - BigWigs comes to mind.  The version numbers are different too, not that we are able to assume to that they are well-formed.
